### ISSUE TITLE
Left function throws IndexOutOfBounds exception

### ISSUE THREAD

**Issue by InverseIntegral** [neo4j/neo4j#12869]
[The documentation](https://neo4j.com/docs/cypher-manual/current/functions/string/#functions-left) of the `left(original, length)` function states that:
> If length exceeds the size of original, original is returned.

However, the following query throws an exception:

```
CREATE ({property: left("OD",3708444074355087135)})
```

```
Exception in thread "main" java.lang.IndexOutOfBoundsException
	at java.base/java.lang.Character.offsetByCodePoints(Character.java:8710)
	at java.base/java.lang.String.offsetByCodePoints(String.java:820)
	at org.neo4j.values.storable.StringWrappingStringValue.substring(StringWrappingStringValue.java:116)
	at org.neo4j.cypher.operations.CypherFunctions.left(CypherFunctions.java:792)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.LeftFunction.compute(StringFunctions.scala:172)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.NullInNullOutExpression.apply(NullInNullOutExpression.scala:32)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.LiteralMap.$anonfun$apply$1(LiteralMap.scala:36)
	at scala.collection.immutable.Map$Map1.foreach(Map.scala:192)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.LiteralMap.apply(LiteralMap.scala:35)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.BaseCreatePipe.setProperties(CreatePipe.scala:53)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.EntityCreatePipe.$anonfun$createNode$2(CreatePipe.scala:94)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.EntityCreatePipe.$anonfun$createNode$2$adapted(CreatePipe.scala:94)
	at scala.Option.foreach(Option.scala:407)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.EntityCreatePipe.createNode(CreatePipe.scala:94)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.CreatePipe.$anonfun$internalCreateResults$2(CreatePipe.scala:138)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.CreatePipe.$anonfun$internalCreateResults$2$adapted(CreatePipe.scala:137)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.CreatePipe.$anonfun$internalCreateResults$1(CreatePipe.scala:137)
	at org.neo4j.cypher.internal.runtime.ClosingIterator$$anon$3.next(ClosingIterator.scala:154)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.EmptyResultPipe.internalCreateResults(EmptyResultPipe.scala:30)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.PipeWithSource.createResults(Pipe.scala:82)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.PipeWithSource.createResults(Pipe.scala:78)
	at org.neo4j.cypher.internal.runtime.interpreted.PipeExecutionResult.request(PipeExecutionResult.scala:69)
	at org.neo4j.cypher.internal.result.StandardInternalExecutionResult.request(StandardInternalExecutionResult.scala:90)
	at org.neo4j.cypher.internal.result.StandardInternalExecutionResult.initiate(StandardInternalExecutionResult.scala:53)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.$anonfun$initiate$1(ClosingExecutionResult.scala:58)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.safely(ClosingExecutionResult.scala:99)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.initiate(ClosingExecutionResult.scala:58)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult$.wrapAndInitiate(ClosingExecutionResult.scala:183)
	at org.neo4j.cypher.internal.CypherCurrentCompiler$CypherExecutableQuery.innerExecute(CypherCurrentCompiler.scala:442)
	at org.neo4j.cypher.internal.CypherCurrentCompiler$CypherExecutableQuery.execute(CypherCurrentCompiler.scala:379)
	at org.neo4j.cypher.internal.ExecutionEngine.doExecute(ExecutionEngine.scala:217)
	at org.neo4j.cypher.internal.ExecutionEngine.$anonfun$executeSubquery$1(ExecutionEngine.scala:178)
	at org.neo4j.cypher.internal.ExecutionEngine.closing(ExecutionEngine.scala:183)
	at org.neo4j.cypher.internal.ExecutionEngine.executeSubquery(ExecutionEngine.scala:175)
	at org.neo4j.cypher.internal.ExecutionEngine.execute(ExecutionEngine.scala:125)
	at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:127)
	at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:115)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:285)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:274)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:268)
```

**Neo4j Version:** 4.4.5
**Operating System:** 5.16.16 Arch
**API:** Java

### Expected behavior
No exception should be thrown and the original string value should be returned.

### Actual behavior
An `IndexOutOfBoundsException` is thrown.

**Comment by InverseIntegral** [comment]
A similar thing happens with the query:
```
CREATE ({property: right("OD", 3708444074355087135)})
```
which throws
```
Exception in thread "main" java.lang.IndexOutOfBoundsException: negative length
	at org.neo4j.cypher.operations.CypherFunctions.right(CypherFunctions.java:882)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.RightFunction.compute(StringFunctions.scala:186)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.NullInNullOutExpression.apply(NullInNullOutExpression.scala:32)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.LiteralMap.$anonfun$apply$1(LiteralMap.scala:36)
	at scala.collection.immutable.Map$Map1.foreach(Map.scala:192)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.LiteralMap.apply(LiteralMap.scala:35)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.BaseCreatePipe.setProperties(CreatePipe.scala:53)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.EntityCreatePipe.$anonfun$createNode$2(CreatePipe.scala:94)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.EntityCreatePipe.$anonfun$createNode$2$adapted(CreatePipe.scala:94)
	at scala.Option.foreach(Option.scala:407)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.EntityCreatePipe.createNode(CreatePipe.scala:94)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.CreatePipe.$anonfun$internalCreateResults$2(CreatePipe.scala:138)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.CreatePipe.$anonfun$internalCreateResults$2$adapted(CreatePipe.scala:137)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.CreatePipe.$anonfun$internalCreateResults$1(CreatePipe.scala:137)
	at org.neo4j.cypher.internal.runtime.ClosingIterator$$anon$3.next(ClosingIterator.scala:154)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.EmptyResultPipe.internalCreateResults(EmptyResultPipe.scala:30)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.PipeWithSource.createResults(Pipe.scala:82)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.PipeWithSource.createResults(Pipe.scala:78)
	at org.neo4j.cypher.internal.runtime.interpreted.PipeExecutionResult.request(PipeExecutionResult.scala:69)
	at org.neo4j.cypher.internal.result.StandardInternalExecutionResult.request(StandardInternalExecutionResult.scala:90)
	at org.neo4j.cypher.internal.result.StandardInternalExecutionResult.initiate(StandardInternalExecutionResult.scala:53)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.$anonfun$initiate$1(ClosingExecutionResult.scala:58)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.safely(ClosingExecutionResult.scala:99)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.initiate(ClosingExecutionResult.scala:58)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult$.wrapAndInitiate(ClosingExecutionResult.scala:183)
	at org.neo4j.cypher.internal.CypherCurrentCompiler$CypherExecutableQuery.innerExecute(CypherCurrentCompiler.scala:442)
	at org.neo4j.cypher.internal.CypherCurrentCompiler$CypherExecutableQuery.execute(CypherCurrentCompiler.scala:379)
	at org.neo4j.cypher.internal.ExecutionEngine.doExecute(ExecutionEngine.scala:217)
	at org.neo4j.cypher.internal.ExecutionEngine.$anonfun$executeSubquery$1(ExecutionEngine.scala:178)
	at org.neo4j.cypher.internal.ExecutionEngine.closing(ExecutionEngine.scala:183)
	at org.neo4j.cypher.internal.ExecutionEngine.executeSubquery(ExecutionEngine.scala:175)
	at org.neo4j.cypher.internal.ExecutionEngine.execute(ExecutionEngine.scala:125)
	at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:127)
	at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:115)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:285)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:274)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:268)
```

**Comment by HannesSandberg** [comment]
Thank you! We will look into this and come back to you.

**Comment by loveleif** [comment]
Thanks again for reporting this @InverseIntegral! We only support lengths in `left` and `right` up to 2^31-1, I've added a more user friendly error in cases like this, this issue will be closed when that PR is merged.

**Comment by loveleif** [comment]
...I changed my mind, we won't error out in cases like this, might as well return the original string as you suggest.

**Comment by InverseIntegral** [comment]
This seems to be fixed in the current version (5.1.0)

**Comment by loveleif** [comment]
Fixed by 079f7d35b9c5b4d9036d92a946c29602e95f4e06 and 15f6b7a57cf5df8d716851b5e8089a3533a020a8
